### PR TITLE
Map TPMI_RH_HIERARCHY type to Esys_TR type

### PR DESCRIPTION
**Pull request by tstruk**
To continue discussion started in https://github.com/tpm2-software/tpm2-tss/pull/1531 the tools can handle the API change by some internal remapping like this.

### CODE REVIEW

File: lib/tpm2.c [hunk 1]
```diff
@@ -591,13 +592,36 @@ tool_rc tpm2_evictcontrol(ESYS_CONTEXT *esys_context,
...
+        case TPM2_RH_ENDORSEMENT:
+            return ESYS_TR_RH_ENDORSEMENT;
+        case ESYS_TR_RH_PLATFORM:
+            return ESYS_TR_RH_PLATFORM;
+        default:
+            return TSS2_ESYS_RC_BAD_VALUE;
+    }
+#elif defined(ESYS_2_3)
+    return hierarchy;
+#else
```
Comment (idesai) [inline lib/tpm2.c]: @tstruk does this branch imply, versions lesser than 2.3? Does it mean also that versions prior to 2.3 did not support ESAPI?

File: lib/tpm2.h [hunk 2]
```diff
@@ -340,4 +340,6 @@ tool_rc tpm2_policy_nv_written(ESYS_CONTEXT *esys_context,
         ESYS_TR policy_session, ESYS_TR shandle1, ESYS_TR shandle2,
         ESYS_TR shandle3, TPMI_YES_NO written_set);
 
+uint32_t fix_esys_hierarchy(TPMI_RH_HIERARCHY hierarchy);
```
Comment (idesai) [inline lib/tpm2.h]: @tstruk is this going to be an errata? if so, should this function be renamed as such with an errata number?

File: lib/tpm2.c (hunk 1 above)
Comment (tstruk) [reply lib/tpm2.c]: It is for tools >= 4.0, which require ESAPI >= 2.3.0.
This is inline with the configure check.

File: lib/tpm2.h (hunk 2 above)
Comment (tstruk) [reply lib/tpm2.h]: My understanding is that the errata mechanism is targeted for TPM spec issues that are fixed by errata. The ESAPI issue that this change fixes is not TPM spec related, but it is the implementation issue, so no errata number exists for it.

Comment (idesai) [reply lib/tpm2.h]: > My understanding is that the errata mechanism is targeted for TPM spec issues that are fixed by errata. The ESAPI issue that this change fixes is not TPM spec related, but it is the implementation issue, so no errata number exists for it.

I see. LGTM.

File: lib/tpm2.c [hunk 3]
```diff
@@ -591,13 +592,36 @@ tool_rc tpm2_evictcontrol(ESYS_CONTEXT *esys_context,
...
+{
+#if defined(ESYS_3_0)
+    switch (hierarchy) {
+        case TPM2_RH_NULL:
+            return ESYS_TR_RH_NULL;
+        case TPM2_RH_OWNER:
+            return ESYS_TR_RH_OWNER;
+        case TPM2_RH_ENDORSEMENT:
+            return ESYS_TR_RH_ENDORSEMENT;
+        case ESYS_TR_RH_PLATFORM:
```
[the remark this answers is not in the document]
Comment (tstruk) [reply lib/tpm2.c]: Yes.

File: lib/tpm2.h (hunk 2 above)
Comment (idesai) [reply lib/tpm2.h]: @tstruk, I think we should leave a comment explaining the necessity for this function and reference the tpm2-tss issue#

Comment (tstruk) [reply lib/tpm2.h]: Agree. Will send an update soon.

File: lib/tpm2.c [hunk 4]
```diff
@@ -591,13 +592,36 @@ tool_rc tpm2_evictcontrol(ESYS_CONTEXT *esys_context,
     return tool_rc_success;
 }
 
+uint32_t fix_esys_hierarchy(TPMI_RH_HIERARCHY hierarchy)
```
Comment (williamcroberts) [inline lib/tpm2.c]: Why a uint32_t and not one of the other typedefs?

Comment (tstruk) [reply lib/tpm2.c]: Because it can return either TPMI_RH_HIERARCHY or ESYS_TR, which are both typedefs of uint32_t anyway.

**Pull request merged**